snuba/pipeline/query_pipeline.py

```diff
@@ -1,44 +1,79 @@
 from abc import ABC, abstractmethod
+from typing import Tuple
 
-
-from snuba.datasets.plans.query_plan import QueryRunner
+from snuba.datasets.plans.query_plan import ClickhouseQueryPlan, QueryRunner
+from snuba.pipeline import Segment
+from snuba.query.logical import Query as LogicalQuery
 from snuba.request import Request
+from snuba.request.request_settings import RequestSettings
 from snuba.web import QueryResult
 
+EntityProcessingPayload = Tuple[LogicalQuery, RequestSettings]
+
+
+class QueryProcessingPipeline(
+    Segment[EntityProcessingPayload, ClickhouseQueryPlan], ABC
+):
+    """
+    A QueryProcessingPipeline contains a series of steps that, given
+    a logical single entity query and request settings, Performs the
+    query processing steps to the point where the query is ready to
+    be executed.
+
+    the query is returned as a ClickhouseQueryPlan.
+
+    TODO: If we will build a processing pipeline for composite query
+    the return type of this pipeline will have to change.
+    """
+
+    @abstractmethod
+    def execute(self, input: EntityProcessingPayload) -> ClickhouseQueryPlan:
+        raise NotImplementedError
 
-class QueryPipeline(ABC):
+
+class QueryExecutionPipeline(Segment[Request, QueryResult], ABC):
     """
-    Contains the instructions to build the query plans for the requested query.
-    The QueryPipeline runs the query plan processors, executes the query plans and
-    returns the relevant result.
+    Contains the instructions to execute a query.
+    The QueryExecutionPipeline performs the all query processing steps and,
+    executes the query plan and returns the result.
 
-    Most of the time, a single query plan is built by the SingleQueryPlanPipeline,
+    Most of the time, a single query plan is built by the SinglePlanExecutionPipeline,
     however under certain circumstances we may have a pipeline with multiple query
     plans. For example, the MultipleQueryPlanPipeline provides a way to build and
     execute more than one query plan and compare their results, which provides a
     way to experiment with different query plans in production without actually
     using their results yet.
 
     This component is produced by the QueryPipelineBuilder.
-
-    TODO: In future, the query pipeline wiil be composed of segments, which will
-    each perform individual processing steps and can be easily reused.
-    Examples of transformations performed by segments are entity processing, query
-    plan building and processing, storage processing, as well as splitters
-    and collectors to help break apart and reassemble joined queries.
     """
 
     @abstractmethod
-    def execute(self) -> QueryResult:
```

Comment:
Actually I ended up reverting this change, since basically all pipeline object are throw away objects that have to be built on the fly.
I think we should revisit this and actually make pipeline just functions, and pass around partially applied functions. There is no need to build an object.